MAINT: Added stacklevel argument to warnings.warn

sumitbinnani opened the pull request:
Updated `numpy/lib/function_base.py`.
Fixes  #5945

sumitbinnani committed:
MAINT: Added ```stacklevel``` argument to ```warnings.warn```

See #5945

jaimefrio commented on numpy/lib/function_base.py:
```diff
@@ -1004,7 +1004,7 @@ def select(condlist, choicelist, default=0):
         warnings.warn("select with an empty condition list is not possible"
                       "and will be deprecated",
-                      DeprecationWarning)
+                      DeprecationWarning, stacklevel = 2)
```
No spaces around the `=` in keyword assignments.

jaimefrio commented on numpy/lib/function_base.py:
```diff
@@ -4206,7 +4209,8 @@ def insert(arr, obj, values, axis=None):
             "in the future the special handling of scalars will be removed "
-            "from insert and raise an error", DeprecationWarning)
+            "from insert and raise an error", DeprecationWarning,
+			stacklevel = 2)
```
I think this should be aligned with the beginning of the string, not with the opening parenthesis.

jaimefrio commented:
Some style nitpicks...

sumitbinnani committed:
STY: Fixed Whitespaces

Fixed indentation to comply with PEP8
Used ```spacecs``` in place of ```tabs```

charris commented:
Umm, we don't allow tabs, use spaces please.

sumitbinnani commented:
@charris  @jaimefrio 
Have fixed indentation issues and used spaces as advised. Kindly review the code and let me know if any more changes are required.

sumitbinnani commented:
Hi @charris @jaimefrio 

Does the changes meet the requirements? Kindly let me know if any additional changes are required.

sumitbinnani commented:
Hi @charris and @jaimefrio,

Is there any issue with the patch? Kindly let me know and would try to submit the patch for the same asap.

njsmith commented:
The changes in the PR look good to me now.

Question 1: the PR summary says that this fixes #5945, but the main question in #5945 is about why a call to `delete` was issuing a warning that referred to `insert` -- does this PR actually address that? (Should it?)

Question 2: any chance you could add some tests for this? I guess adding tests for all the little deprecation warning tweaks would be tiresome and of limited value, because the code will just get removed again in a few versions anyway. But for the `RuntimeWarning`s in `cov`, `_median`, and `_percentile` it seems particularly useful, because the `RuntimeWarning`s will be around indefinitely, and because it's not 100% obvious whether stacklevel=2 is even correct for `_median` and `_percentile` which are internal helper functions. (It might be because I think the code the callers are written in C and thus don't affect the stacklevel= count, but it's not 100% obvious.) The way to do this would be to use the [`catch_warnings`](https://docs.python.org/2/library/warnings.html#testing-warnings) context manager with `record=True`, call the function that should issue a warning, and then check that the `.filename` attribute of the recorded warning object refers to the file that called the function rather than some internal numpy file.

Sorry for the slow response, and thanks for being persistent in pinging us!

sumitbinnani commented:
Thanks @njsmith for the inputs. I will look into the details during weekend and update the thread.

homu commented:
:umbrella: The latest upstream changes (presumably #7181) made this pull request unmergeable. Please resolve the merge conflicts.

sumitbinnani commented:
Can you elaborate on what needs to be done, @homu?

rkern commented:
@homu is a bot and will not reply. You will need to manually merge in `master` and fix the conflicts.

charris commented:
This has been taken care of in #7148 along with a number of other cases, so closing this. Thanks @sumitbinnani, I apologize that we didn't get this in earlier.
